atuin crypto blog post (#1079)
ellie/atuin: docs/blog/2023/07/07-new-encryption/index.md

---
title: Atuin's new encryption scheme
description: Details about the new encryption system for Atuin
slug: new-encryption
authors: [conrad]
tags: [insights]
---

End-to-end encryption is an essential component of Atuin.

> One of our core philosophies, when we created the sync service, is that
> we want no reason to worry about user data. The shell is a very

Reviewer: ```suggestion
One of our core philosophies, when we created the sync service, was that
we didn't want to worry about storing user data. The shell is a very
```

### We use the same key for each message

Shell history is quite predictable. If you have a 2-byte history entry, it's quite likely that it's `ls`. Given the encrypted blob, you can start to brute force the associated key. A proof was published stating that no attack on Salsa20 with 128-bit key is possible with an average search time of less than 2^130 (about 10^39) random guesses.

To put that number into perspective. Performing 1 billion key operations per CPU core per second, and using a suite of 1 billion CPU cores, the attack will take roughly 10 trillion years.

Atuin uses a 256-bit key which is even more secure, and therefore not at risk of a practical brute-force attack. It follows that we are likely safe from a known plain-text attack.

However, there is still the issue of key leaking. We have no key-upgrade policy.
If a key is leaked, maybe through a side-channel attack, a social attack, or malware,
then the only solution is to create a new account with a new key.

This is partially an issue.

## Change

> While researching these systems, I have learnt a lot more new techniques that modern cryptographic systems use. While the analysis above indicates that we are protected, there might be attacks we are unaware of, so keeping up with modern research is important.

Reviewer: ```suggestion
While researching these systems, I learnt of many new cryptographic techniques that some modern systems use. While the analysis above indicates that we are protected, there might be attacks we are unaware of, so keeping up with modern research is important.
```

We're also in the middle of redesigning our sync service. While we're already planning a big change, we might as well consider updating the encryption too

### Key wrapping

A common approach to encrypting lots of different data is the use of wrapped keys.
The idea here is that each payload has an ephemeral encryption key.
This is then encrypted (wrapped) using the master key and stored with the data.

Initially, this seemed less secure to me. However, my analysis seems to point out that
the master key is less vulnerable to side-channel attacks since it is less used.
It also offers no decrease in security since brute-forcing the master key from an
ephemeral data key is just as hard as it is for any message. In the end, it's like
a password manager for your encrypted data.

> This unlocks some potential future upgrades.

Reviewer: ```suggestion
This would unlock some potential future upgrades.
```

### Stronger ciphers

XSalsa20 was later superseded by XChaCha20 by the same author. It has a very similar construction,
but the stream cipher has better mixing characteristics, which makes any non-brute-force attacks harder to craft.

## Conclusion

I started to craft a new solution using these concepts.
But eventually, I realised that I shouldn't be reinventing the wheel here.
During more and more of my research sessions, I stumbled upon [PASETO](https://paseto.io/).
While the intended use case is security tokens,
their local encryption scheme is designed such that encrypted data is safe to be shared publically.
Their V4 scheme also uses the XChaCha20 cipher which I was initially planning to use.

In the end, I bit the bullet and decided to use the standard.
The nice thing with secretbox is that existing implementations in other languages are widely available.
Making it easy to implement sync in third parties. If we implemented our own scheme,
that would make it much easier for third parties to make mistakes if they wanted to use the sync data directly.

Using PASETO, there are existing implementations that we didn't have to write.

> This means that we don't build software doomed to die a lonely death. It also means

Reviewer: :rofl:
